Fix docs multiple nested and multiple methods

Woile opened the pull request:
# Description
Multiple nested resources with multiple methods were not displayed. Also
`detail_routes` and `list_routes` with multiple methods. This PR attempts to fix it. 
I created a template tag, that **recursively** finds the links inside a given section. What do you guys think?
I am not sure if the format I've given to the generated links it's okay, but it's a start.

Issues related:
#5324 
[#4391](https://github.com/encode/django-rest-framework/issues/4391)
#4965 

## Results

Default actions
```
users: {
    list()
    create()
    read(id)
    update(id)
}
```    
![screenshot from 2017-08-15 16-18-56](https://user-images.githubusercontent.com/2585816/29321988-5b56bc4a-81dc-11e7-9d58-984226b55ee6.png)

--------------------------------
Single action
```
users: {
    list([page])
}
```
![screenshot from 2017-08-15 16-19-30](https://user-images.githubusercontent.com/2585816/29321968-52204a88-81dc-11e7-949a-d39d560a8543.png)

--------------------------------
Single detail_route url
```
users: {
    list()
    create()
    read(id)
    update(id)
    farmers(id) <--
}
```
![screenshot from 2017-08-15 16-20-46](https://user-images.githubusercontent.com/2585816/29322021-65f9a766-81dc-11e7-8e68-c271d498109f.png)

--------------------------------
More than one detail_route url
```
users: {
    farmers: {
        read(id) <--
        update(id) <--
    }
    list()
    create()
    read(id)
    update(id)
}
```
![screenshot from 2017-08-15 16-22-17](https://user-images.githubusercontent.com/2585816/29322028-70d9a65e-81dc-11e7-979f-4945c47ec758.png)

--------------------------------
Nested routes
```
users: {
        animals: {
            dogs: {
                list(user_id, [page]) <--
                create(user_id, email, content, created) <--
            }
            list(user_id, [page]) <--
            create(user_id, email, content, created) <--
        }
        farmers: {
            read(id)
            update(id)
        }
        list()
        create()
        read(id)
        update(id)
    }
```
![screenshot from 2017-08-15 16-23-16](https://user-images.githubusercontent.com/2585816/29322050-8288c1e6-81dc-11e7-8d89-b1755b1f6c6e.png)

--------------------------------
Multiple resources with nested schema
```
users: {
    animals: {
        dogs: {
            list(user_id, [page])
            create(user_id, email, content, created)
        }
        list(user_id, [page])
        create(user_id, email, content, created)
    }
    farmers: {
        read(id)
        update(id)
    }
    list()
    create()
    read(id)
    update(id)
}
vets: {  <--
    animals: {
        dogs: {
            list(user_id, [page])
            create(user_id, email, content, created) 
        }
    }
}
```
![screenshot from 2017-08-15 16-24-25](https://user-images.githubusercontent.com/2585816/29322078-9989c53e-81dc-11e7-8bd0-26c51a38c75a.png)

--------------------------------
Multiple list_route url
```
users: {
    farmers: {
        read(id)
        update(id)
    }
    houses: {
        read() <--
        create() <--
    }
    list()
    create()
    read(id)
    update(id)
}
```
![screenshot from 2017-08-15 16-31-20](https://user-images.githubusercontent.com/2585816/29322105-a9c78b20-81dc-11e7-9489-e5a04aa89385.png)

--------------------------------
Single list_route url (houses)
```
users: {
    farmers: {
        read(id)
        update(id)
    }
    list()
    create()
    houses() <--
    read(id)
    update(id)
}
```
![screenshot from 2017-08-15 16-32-25](https://user-images.githubusercontent.com/2585816/29322111-afa7ff66-81dc-11e7-8544-178bebbcae3f.png)

carltongibson commented:
This is looking good. Thanks!

I'll review fully this week.

carltongibson commented on rest_framework/templatetags/rest_framework.py:
```diff
@@ -245,6 +245,29 @@ def items(value):
+    """
+    Recursively find every link in a schema, even nested.
+    """
+    NESTED_FORMAT = '%s > %s'
```
Is this going to be valid when used as an HTML `id` attribute?

Woile replied:
works fine in firefox and chrome under debian x64

e.g:
`http://localhost:8000/docs/#users-animals > dogs > list`

also clicking the links on the sidebar work.

carltongibson replied:
What about for the JavaScript data binding in https://github.com/encode/django-rest-framework/blob/master/rest_framework/templates/rest_framework/docs/interact.html#L19 ?

carltongibson commented:
@Woile I think this is great! 💃🏼

I like the `Vets > animals > dogs > list` example. It suggests the CSS is robust 
enough. _Flat is better than nested_: If you need more nesting than that you'll 
need to modify the CSS for yourself. (This is nice limit to draw.)

**Tests**: The examples you've give seem to cover the core cases. Can you turn them into 
unit tests?

carltongibson commented:
My only question is (the above) about the JavaScript data-binding — does that break? ([I found it did working around this (comment) in a worse way](https://github.com/encode/django-rest-framework/issues/5077#issuecomment-302104652) in #5077)

If so returning an `id` ≈`nested-formated-string` and then using another template tag for display might be enough...

carltongibson commented:
[HTML 5 Spec says of ids](https://www.w3.org/TR/html5/dom.html#the-id-attribute): 

>  The value must not contain any space characters.

I'm sure it'll work for some clients but break in others.

Not conclusive but my mail client makes a mess of it: 

<img width="674" alt="screenshot 2017-08-16 21 41 39" src="https://user-images.githubusercontent.com/64686/29382076-c35caf9e-82cb-11e7-85dd-28d4215a4dca.png">

I think we should adjust this for robustness sake. 🙂

Woile commented:
@carltongibson for the ID's maybe we could use the [slugify](https://docs.djangoproject.com/en/1.11/ref/utils/#django.utils.text.slugify) from django which will convert spaces to hiphens. That'd be a solution to deal with the id related problems, right?

I'll try to make the unittests tomorrows.

carltongibson commented:
> I'll try to make the unittests tomorrow.

Great. Thanks! 

I think tests at the `schema_links` function level will be fine — i.e. just passing a simplified schema document and checking the result; we should be able to trust the whole `register.filter` and rendering steps, which aren't so nice to test anyway.

Woile commented:
@carltongibson I fixed the issues you suggested, tell me what you think. 

The generated ID for links now looks like this:
http://localhost:8000/docs/#users-objects-list

And the interaction modal works fine (note: I added a normalize function over the `data-key` in the form).

I'll be working on the tests now.

carltongibson commented:
@Woile — Looking AWESOME! Thanks for the effort!!! 

(If we can get the tests in place, I think we can get this in.)

Woile commented:
@carltongibson I've added the tests, tell me what you think. They are a bit long, but the cases are long as well :P

carltongibson commented:
@Woile The test cases look great. I'll give this a final review today. 

Thank you for you effort!

carltongibson commented:
@Woile This is good work: it passes my manual testing. 🙂 I'm going to pull this in. Thanks again!

Woile commented:
@carltongibson you are welcome, no problem at all! Glad to help
